_source/_posts/2022-09-19-kafka-microservices.md

```diff
@@ -0,0 +1,772 @@
+
+```java
+package com.okta.developer.alert.service;
+
+import com.okta.developer.alert.config.KafkaStoreAlertConsumer;
+import com.okta.developer.alert.domain.StoreAlert;
+import com.okta.developer.alert.repository.StoreAlertRepository;
+import com.okta.developer.alert.service.dto.StoreAlertDTO;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.cloud.stream.annotation.StreamListener;
+import org.springframework.messaging.Message;
+import org.springframework.stereotype.Service;
+
+import java.time.Instant;
+@Service
+public class AlertConsumer {
+
+    private final Logger log = LoggerFactory.getLogger(AlertConsumer.class);
+
+    private StoreAlertRepository storeAlertRepository;
+
+    private EmailService emailService;
+
+    public AlertConsumer(StoreAlertRepository storeAlertRepository, EmailService emailService) {
+        this.storeAlertRepository = storeAlertRepository;
+        this.emailService = emailService;
+    }
+
+    @StreamListener(value = KafkaStoreAlertConsumer.CHANNELNAME, copyHeaders = "false")
+    public void consume(Message<StoreAlertDTO> message) {
+        StoreAlertDTO dto = message.getPayload();
+        log.info("Got message from kafka stream: {} {}", dto.getStoreName(), dto.getStoreStatus());
+        try {
+            StoreAlert storeAlert = new StoreAlert();
+            storeAlert.setStoreName(dto.getStoreName());
+            storeAlert.setStoreStatus(dto.getStoreStatus());
+            storeAlert.setTimestamp(Instant.now());
+
+            storeAlertRepository.save(storeAlert);
+            emailService.sendSimpleMessage(dto);
+        } catch (Exception e) {
+            log.error(e.getMessage(), e);
+        }
+    }
+}
+```
+
+**NOTE**: Any unhandled exception during message processing will make the service leave the consumer group. That's why there's code above that catches `Exception`.
+
+As a last customization step, update the logging configuration the same way you did for the `store` microservice.
+
+## Microservices + Kafka Container Deployment
+
+Modify `docker-compose/docker-compose.yml` and add the following environment variables for the `alert` application:
+
+```yaml
+- SPRING_MAIL_USERNAME=${MAIL_USERNAME}
+- SPRING_MAIL_PASSWORD=${MAIL_PASSWORD}
+- ALERT_DISTRIBUTION_LIST=${DISTRIBUTION_LIST}
+```
+
+Edit `docker-compose/.env` and add values for the new environment variables:
+
+```bash
+MAIL_USERNAME={yourGmailAccount}
+MAIL_PASSWORD={yourPassword}
+DISTRIBUTION_LIST={anotherEmailAccount}
+```
+
+Make sure Docker Desktop is running, then generate the Docker image for the `store` microservice. Run the following command from the `store` directory.
+
+```bash
+./mvnw -ntp -Pprod verify jib:dockerBuild
+```
+
+Repeat for the `alert` and `gateway` apps.
```

Comment:
I get an error when I try running this command in the gateway app:

```
[INFO] - Generating browser application bundles (phase: setup)...
[INFO] An unhandled exception occurred: Cannot find module 'caniuse-lite/data/features/css-unicode-bidi'
[INFO] Require stack:
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/autoprefixer/data/prefixes.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/autoprefixer/lib/autoprefixer.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/postcss-preset-env/dist/index.cjs
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular-devkit/build-angular/src/webpack/configs/styles.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular-devkit/build-angular/src/webpack/configs/index.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular-devkit/build-angular/src/builders/browser/index.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular-devkit/build-angular/src/index.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular-builders/custom-webpack/dist/browser/index.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular/cli/node_modules/@angular-devkit/architect/node/node-modules-architect-host.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular/cli/node_modules/@angular-devkit/architect/node/index.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular/cli/src/command-builder/architect-base-command-module.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular/cli/src/command-builder/architect-command-module.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular/cli/src/commands/build/cli.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular/cli/src/command-builder/command-runner.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular/cli/lib/cli/index.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular/cli/lib/init.js
[INFO] See "/private/var/folders/f7/5x4rvsr5599ffnfckn487nzm0000gn/T/ng-8UMHEM/angular-errors.log" for further details.
```

I'm guessing it's because I'm on airplane wi-fi right now. I'll keep trying.

Reply:
It finally worked.